io/shared/src/main/scala/fs2/io/net/tls/TLSSocket.scala

```diff
@@ -38,9 +38,19 @@ sealed trait TLSSocket[F[_]] extends Socket[F] with TLSSocketPlatform[F] {
   def session: F[SSLSession]
 
   /** Provides access to the current application protocol that has been negotiated.
+    *
+    * Raises a [[NoSuchElementException]] if it has not yet been determined if application protocols might
+    * be used for this connection. See [[applicationProtocolOption]] for a safer option.
+    *
+    * @see https://discord.com/channels/632277896739946517/632310980449402880/1100649913223745597
     */
   def applicationProtocol: F[String]
```

Comment:
I think we can even deprecate this API to encourage the other one :)

Not sure about linking to Discord. We should record anything important either in discussion on this PR or in code / comments themselves.

Reply:
In my experience linking to Discord gives the perspective on back-and-forth that we had until we arrived at this. It's not super important, but it might give some insight into the context for a person looking at it.